README.md
## Bean Instagram

[Bean Instagram](http://themebeans.com/plugin/bean-instagram-plugin/) is a WordPress plugin by [ThemeBeans](http://themebeans.com) that adds an Instagram widget to _Appearence > Widgets_.

The widget can display your Instagram feed, recent uploads, or liked shots.

For more information, check out the [plugin page](http://themebeans.com/plugin/bean-instagram-plugin/).


## How to Install

1. Copy the plugin folder 'bean-instagram' to the '/wp-content/plugins' directory of your WordPress installation.
2. Go to the _WordPress Dashboard > Plugins_ screen and click _Activate_ under _Bean Instagram_.


## How to Use

1. Go to the _WordPress Dashboard > Appearance > Widgets_ screen.
2. Drag/drop the _Bean Instagram_ widget to any of the widget areas.
3. Follow [this](http://themebeans.com/registering-your-instagram-app-to-retrieve-your-client-id-secret-code/) guide to retrieve your app codes.


changelog.txt
SEP 1 2015 - Version 2.0.1
<ul>
<li>WP4.3 _construct update</li>
</ul>

SEP 11 2014 - Version 2.0
<ul>
<li>OAuth update</li>
<li>Moved OAuth activation from widget to Settings > Bean Instagram</li>
<li>Added "Settings" link on plugins.php</li>
<li>Performance Improvements</li>
</ul>

APR 24 2014 - Version 1.3.1
<ul>
<li>Minor UI Adjustments for WP3.9</li>
</ul>

FEB 12 2014 - Version 1.3
<ul>
<li>Added support for auto activation & updates</li>
</ul>

DEC 17 2013 - Version 1.2
<ul>
<li>Performance Improvements</li>
</ul>

AUG 12 2013 - Version 1.1
<ul>
<li>Performance Improvements</li>
</ul>

JULY 3 2013 - Version 1.0
<ul>
<li>Initial release</li>
</ul>

bean-instagram.php
<?php
/*
Plugin Name: Bean Instagram
Plugin URI: http://themebeans.com/plugins/bean-instagram
Description: Enables an Instagram feed widget. You must register an <a href="http://instagram.com/developer/" target="_blank">Instagram App</a> to retrieve your client ID and secret. <a href="http://themebeans.com/tutorials/retrieving-your-instagram-id-secret/">Learn More</a>
Version: 2.1
Author: ThemeBeans
Author URI: http://themebeans.com
*/

// DON'T CALL ANYTHING
if ( !function_exists( 'add_action' ) ) {
	echo 'Hi there!  I\'m just a plugin, not much I can do when called directly.';
	exit;
}

define('BEAN_INSTAGRAM_PATH', plugin_dir_url( __FILE__ ));
define('BEAN_INSTAGRAM_PLUGIN_FILE', __FILE__ );


/*===================================================================*/
/*
/* PLUGIN FEATURES SETUP
/*
/*===================================================================*/
$bean_plugin_features[ plugin_basename( __FILE__ ) ] = array(
        "updates"  => false // Whether to utilize plugin updates feature or not
    );


if ( ! function_exists( 'bean_plugin_supports' ) ) 
{
    function bean_plugin_supports( $plugin_basename, $feature ) 
    {
        global $bean_plugin_features;

        $setup = $bean_plugin_features;

        if( isset( $setup[$plugin_basename][$feature] ) && $setup[$plugin_basename][$feature] )
            return true;
        else
            return false;
    }
}




/*===================================================================*/
/*
/* PLUGIN UPDATER FUNCTIONALITY
/*
/*===================================================================*/
define( 'EDD_BEANINSTAGRAM_TB_URL', 'http://themebeans.com' );
define( 'EDD_BEANINSTAGRAM_NAME', 'Bean Instagram' );

if ( bean_plugin_supports ( plugin_basename( __FILE__ ), 'updates' ) ) : // check to see if updates are allowed; only import if so

//LOAD UPDATER CLASS
if( !class_exists( 'EDD_SL_Plugin_Updater' ) ) 
{
    include( dirname( __FILE__ ) . '/updates/EDD_SL_Plugin_Updater.php' );
}
//INCLUDE UPDATER SETUP
include( dirname( __FILE__ ) . '/updates/EDD_SL_Activation.php' );


endif; // END if ( bean_plugin_supports ( plugin_basename( __FILE__ ), 'updates' ) )




/*===================================================================*/
/* UPDATER SETUP
/*===================================================================*/
function beaninstagram_license_setup() 
{
	add_option( 'edd_beaninstagram_activate_license', 'BEANINSTAGRAM' );
	add_option( 'edd_beaninstagram_license_status' );
}
add_action( 'init', 'beaninstagram_license_setup' );

function edd_beaninstagram_plugin_updater() 
{
    // check to see if updates are allowed; don't do anything if not
    if ( ! bean_plugin_supports ( plugin_basename( __FILE__ ), 'updates' ) ) return;

	//RETRIEVE LICENSE KEY
	$license_key = trim( get_option( 'edd_beaninstagram_activate_license' ) );

	$edd_updater = new EDD_SL_Plugin_Updater( EDD_BEANINSTAGRAM_TB_URL, __FILE__, array( 
			'version' => '2.0',
			'license' => $license_key,
			'item_name' => EDD_BEANINSTAGRAM_NAME,
			'author' 	=> 'ThemeBeans'
		)
	);
}
add_action( 'admin_init', 'edd_beaninstagram_plugin_updater' );




/*===================================================================*/
/* ADD SETTINGS LINK TO PLUGINS PAGE
/*===================================================================*/
define( 'BEANINSTAGRAM_PLUGIN_BASENAME', plugin_basename( __FILE__ ) );

add_filter( 'plugin_action_links', 'beaninstagram_plugin_action_links', 10, 2 );

function beaninstagram_plugin_action_links( $links, $file ) {
	if ( $file != BEANINSTAGRAM_PLUGIN_BASENAME )
		return $links;

	$settings_link = '<a href="' . menu_page_url( 'bean-instagram', false ) . '">'
		. esc_html( __( 'Settings', 'bean-instagram' ) ) . '</a>';

	array_unshift( $links, $settings_link );

	return $links;
}




/*===================================================================*/
/* DEACTIVATION HOOK - REMOVE OPTION
/*===================================================================*/
function beaninstagram_deactivate() 
{
	delete_option( 'edd_beaninstagram_activate_license' );
	delete_option( 'edd_beaninstagram_license_status' );
}
register_deactivation_hook( __FILE__, 'beaninstagram_deactivate' );




/*===================================================================*/
/* BEGIN BEAN INSTAGRAM PLUGIN
/*===================================================================*/
require_once('bean-instagram-widget.php');
?>

js/bean-instagram.js
function clear_accesstoken(error_message) {
    document.getElementById("bean_inst_access_token").value = "";
    document.getElementById("bean_inst_plugin_userid").value = "";

    if (document.getElementById("bean_inst_access_token_status")) {
        document.getElementById("bean_inst_access_token_status").style.color = "red";
        document.getElementById("bean_inst_access_token_status").innerHTML = error_message;
    }
}

function bean_instagram_get_access_token(redirect_uri) {
    var clientid = document.forms[0].elements['bean_inst_client_id'].value;

    if (clientid.length == 0) {
        document.forms[0].elements['bean_inst_client_id'].focus();

        if (!document.getElementById("bean_inst_client_id_error")) {
            var pNode = document.createElement("strong");
            pNode.innerHTML = "Please fill in the \"Client ID\" field above first!";
            pNode.setAttribute('id', 'bean_inst_client_id_error');
            pNode.style.paddingLeft = "10px";
            pNode.style.lineHeight = "28px";
            this.parentNode.appendChild(pNode);
        }

        return;
    }

    var form = document.createElement("form");
    form.setAttribute('id', 'instagram_auth');
    form.setAttribute('name', 'instagram_auth');
    form.setAttribute('action', 'https://api.instagram.com/oauth/authorize/');
    form.setAttribute('method', 'GET');

    var responseType = document.createElement('input');
    responseType.setAttribute('type', 'hidden');
    responseType.setAttribute('name', 'response_type');
    responseType.setAttribute('value', 'code');
    responseType.setAttribute('id', 'instagram_auth_response_type');

    var redirectURI = document.createElement('input');
    redirectURI.setAttribute('type', 'hidden');
    redirectURI.setAttribute('name', 'redirect_uri');
    redirectURI.setAttribute('value', redirect_uri);
    redirectURI.setAttribute('id', 'instagram_auth_redirect_uri');

    var clientID = document.createElement('input');
    clientID.setAttribute('type', 'hidden');
    clientID.setAttribute('name', 'client_id');
    clientID.setAttribute('value', clientid);
    clientID.setAttribute('id', 'instagram_auth_client_id');

    form.appendChild(responseType);
    form.appendChild(redirectURI);
    form.appendChild(clientID);

    form.submit();

    event.preventDefault();
    return false;
}

updates/EDD_SL_Activation.php
<?php
/**
 * This file sets up the auto license activater.
 *
 *
 * @package Bean Plugins
 * @subpackage BeanSocial
 * @author ThemeBeans
 * @since BeanSocial 1.4
 */
 
 
/*===================================================================*/
/* ACTIVATION
/*===================================================================*/
function edd_beaninstagram_activate_license() 
{
	//GET STATUS
	$status = get_option( 'edd_beaninstagram_license_status' );
	
	//RETRIEVE LICENSE KEY
	$license = trim( get_option( 'edd_beaninstagram_activate_license' ) );
	
	//UNCOMMENT TO TEST LICENSE INPUT	
	//echo 'License'; echo $license; echo ' / Status'; echo $status;
	
	//CHECK IF STATUS IS INVALID
	if( $status == '' OR $status == 'invalid' ) 
	{
		//DATA TO SEND WITH OUR API REQUEST
		$api_params = array( 
			'edd_action'=> 'activate_license', 
			'license' 	=> $license, 
			'item_name' => urlencode( EDD_BEANINSTAGRAM_NAME )
		);
		
		//CALL CUSTOM API
		$response = wp_remote_get( add_query_arg( $api_params, EDD_BEANINSTAGRAM_TB_URL ), array( 'timeout' => 15, 'sslverify' => false ) );
	
		//MAKE SURE RESPONSE IS OK
		if ( is_wp_error( $response ) )
			return false;
	
		//DECODE DATA
		$license_data = json_decode( wp_remote_retrieve_body( $response ) );
		
		//$license_data->license will be either "active" or "inactive"
	
		update_option( 'edd_beaninstagram_license_status', $license_data->license );
	}
}
add_action('admin_init', 'edd_beaninstagram_activate_license');

bean-instagram-widget.php
<?php
/**
 * Widget Name: Bean Instagram Widget
 */

// Register widget
add_action('widgets_init', create_function('', 'return register_widget("Bean_Instagram_Widget");'));

class Bean_Instagram_Widget extends WP_Widget 
{
	// Constructor
	function __construct() {
		parent::__construct(
			'bean_instagram', // Base ID
			__( 'Bean Instagram', 'bean' ), // Name
			array( 'description' => __( 'Displays your Instagram feed.', 'bean' ), ) // Args
		);

		if ( is_active_widget(false, false, $this->id_base) ) {
	        add_action( 'wp_head', array(&$this, 'load_widget_style') );
	     } 
	}


	/*===================================================================*/
	/*	LOAD WIDGET STYLE DEPENDING ON THE THEME ( OR DEFAULT IF NO ASSOCIATED STYLE WAS FOUND )
	/*===================================================================*/

	public function load_widget_style() {
		$current_theme = wp_get_theme();

		$theme_css_path = DIRNAME(__FILE__) . '/themes/' . $current_theme . '/instagram.css';
		$theme_css_url = plugin_dir_url(__FILE__) . 'themes/' . $current_theme . '/instagram.css';
		$default_css_url = plugin_dir_url(__FILE__) . 'themes/_default/instagram.css';

		// fix spaces that may exist in the theme name
		$theme_css_url = str_replace(' ', '%20', $theme_css_url);

		if (file_exists($theme_css_path))
			wp_enqueue_style( 'bean-instagram-style', $theme_css_url, false, '1.0', 'all' );
		else
			wp_enqueue_style( 'bean-instagram-style', $default_css_url, false, '1.0', 'all' );
	}

	/*===================================================================*/
	/*	DISPLAY WIDGET
	/*===================================================================*/
	public function widget($args, $instance) {
		extract($args);
		if(!empty($instance['title'])){ $title = apply_filters( 'widget_title', $instance['title'] ); }

		$desc = $instance['desc'];

		echo $before_widget;
		if ( ! empty( $title ) ){ echo $before_title . $title . $after_title; }

		if($desc != '') : ?><p><?php echo $desc; ?></p><?php endif;

		$accesstoken = self::getAccessToken();

		// CHECK SETTINGS & DIE IF NOT SET
		if(!$accesstoken || empty($instance['cachetime'])){
			echo '<strong>Please fill all the widget settings under "Settings > Bean Instagram" and request an access token!</strong>' . $after_widget;
			return;
		}

		// CHECK IF CACHE NEEDS UPDATE
		$ip_instagram_plugin_last_cache_time = self::getLastCacheTime();
		$diff = time() - $ip_instagram_plugin_last_cache_time;
		$crt = $instance['cachetime'] * 3600;

		//	YUP, NEEDS ONE
		if($diff >= $crt || empty($ip_instagram_plugin_last_cache_time)) {

			// PREPARE THE ENDPOINT URI DEPENDING ON THE TYPE OF CONTENT TO SHOW
			switch($instance['show']) {
				case 'recent' :
								// GET THE USER ID
				$user_id = self::getUserId();

				$endpoint_uri = "https://api.instagram.com/v1/users/" . $user_id . "/media/recent/?count=8&access_token=" . $accesstoken;
				break;

				case 'liked' :
				$endpoint_uri = "https://api.instagram.com/v1/users/self/media/liked?count=8&access_token=" . $accesstoken;
				break;

				default :
				$endpoint_uri = "https://api.instagram.com/v1/users/self/feed?count=8&access_token=" . $accesstoken;
				break;
			}

			// WE HAVE AN ENDPOINT_URI SET
			if (isset($endpoint_uri) && !empty($endpoint_uri)) {
				$ch = curl_init();

				curl_setopt($ch, CURLOPT_URL, $endpoint_uri);
				curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

				$result = curl_exec($ch);

				curl_close($ch);

				// PARSE THE RESULT TO MAKE IT USEABLE
				$result = json_decode($result);

				// CHECK IF THE RESULT HAS SOME ERROR
				if (@isset($result->meta->error_type)) {

					$cache_error = 1;

					// CHECK IF THERE IS A PROBLEM IN THE ACCESS TOKEN (EITHER INVALID OR EXPIRED)
					if ($result->meta->error_type == 'OAuthAccessTokenError'
						|| $result->meta->error_type == 'OAuthAccessTokenException') {

						// DELETE THE ACCESS TOKEN
						self::deleteAccessToken();

					$cache_error = 2;
				}
			}
			else {
					// GO THROUGH THE RESULT, EXTRACT THE USEFUL INFORMATION, AND STORE IN THE ARRAY
				$posts_array = array();

				if (@isset($result->data))
					foreach($result->data as $post) {
						if (!is_object($post)) continue;

						$post_entry = array(
							"caption" => $post->caption == "null" ? "" : preg_replace('/[^(\x20-\x7F)]*/','', isset($post->caption->text) ? $post->caption->text : ''),
							"link" => $post->link,
							"image" => $post->images->thumbnail->url
							);

						array_push($posts_array, $post_entry);
					}
				}

			}

			// SAVE POSTS TO WP OPTION
			if (isset($posts_array)) {
				self::setPostsCache($posts_array);
				self::setLastCacheTime(time());
			}

			echo '<!-- instagram cache has been updated! -->';
		}

		// DISPLAY ERROR MESSAGE IF THE ACCESS TOKEN SEEMS TO BE INVALID OR HAS EXPIRED
		if (@isset($cache_error)) {
			?>
			<p>Error while fetching cache:
				<?php
				switch($cache_error) {
					case 1 : echo "An undefined error occurred while contacting the Instagram API.";
					break;

					case 2 : echo "The access token seems to be invalid or has expired.";
					break;
				}?>

			</p>

			<?php
		}

		$ip_instagram_plugin_posts = self::getPostsCache();
		$ip_instagram_plugin_posts = maybe_unserialize($ip_instagram_plugin_posts);

		if (!empty($ip_instagram_plugin_posts)) {
			?>

			<div class="instagram-image-wrapper">
				<?php
				$i = 0;
				foreach($ip_instagram_plugin_posts as $post) {
					?>

					<div class="instagram_badge_image" id="instagram_badge_image<?php echo $i; ?>">
						<a href="<?php echo $post['link']; ?>" target="_blank">
							<img src="<?php echo $post['image']; ?>" alt="A post on Instagram" title="<?php echo $post['caption']; ?>">
						</a>
					</div>

					<?php

					$i++;
				}

				?>
			</div>

			<?php
		} else {
			?>
			<strong>Could not retrieve Instagram posts.</strong>
			<?php
		}

		echo $after_widget;
	}


	/*===================================================================*/
	/*	UPDATE WIDGET
	/*===================================================================*/
	public function update($new_instance, $old_instance) {
		$instance = array();
		$instance['title'] = strip_tags( $new_instance['title'] );
		$instance['desc'] = stripslashes( $new_instance['desc'] );
		$instance['cachetime'] = strip_tags( $new_instance['cachetime'] );
		$instance['show'] = strip_tags( $new_instance['show'] );

		if($old_instance['show'] != $new_instance['show']
			|| $old_instance['cachetime'] != $new_instance['cachetime'] ) {

			// SET THE LAST CACHE TIME TO ZERO SO THAT THE POSTS ARE RE-CACHED
			self::setLastCacheTime(0);
	}

	return $instance;
}


/*===================================================================*/
	/*	WIDGET SETTINGS (FRONT END PANEL)
	/*===================================================================*/
	public function form($instance) {
		$defaults = array( 'title' => 'Bean Instagram Plugin',
			'desc'  => '',
			'cachetime' => '5',
			'show' => 'recent'
			);

		$instance = wp_parse_args( (array) $instance, $defaults );


		?>

		<p>
			<label for="<?php echo $this->get_field_id( 'title' ); ?>"><?php _e( 'Title:', 'bean' ); ?></label>
			<input class="widefat" id="<?php echo $this->get_field_id( 'title' ); ?>" name="<?php echo $this->get_field_name( 'title' ); ?>" type="text" value="<?php echo esc_attr( $instance['title'] ); ?>" />
		</p>

		<p style="margin-top: -8px;">
			<textarea class="widefat" rows="5" cols="15" id="<?php echo $this->get_field_id( 'desc' ); ?>" name="<?php echo $this->get_field_name( 'desc' ); ?>"><?php echo $instance['desc']; ?></textarea>
		</p>

		<p>
			<label for="<?php echo $this->get_field_id( 'show' ); ?>"><?php _e('Show:', 'bean'); ?></label>
			<select id="<?php echo $this->get_field_id( 'show' ); ?>" name="<?php echo $this->get_field_name( 'show' ); ?>" class="widefat">
				<option value='recent' <?php if ( 'recent' == $instance['show'] ) echo 'selected="selected"'; ?>>Recent Uploads</option>
				<option value='feed' <?php if ( 'feed' == $instance['show'] ) echo 'selected="selected"'; ?>>Following</option>
				<option value='liked' <?php if ( 'liked' == $instance['show'] ) echo 'selected="selected"'; ?>>Liked Posts</option>
			</select>
		</p>

		<p><label for="<?php echo $this->get_field_id( 'cachetime' ); ?>"><?php _e( 'Cache every:', 'bean' ); ?></label>
			<input type="text" name="<?php echo $this->get_field_name( 'cachetime' ); ?>" id="<?php echo $this->get_field_id( 'cachetime' ); ?>" value="<?php echo esc_attr($instance['cachetime']) ?>" class="small-text" /> hours</p>

			<?php
		}

     /*===================================================================*/
     /*  SUPPORT FUNCTIONS TO HANDLE THE INSTAGRAM ACCESS TOKEN
     /*===================================================================*/

     private function getAccessToken() {
     return get_option('bean_inst_access_token');
     }

     /*===================================================================*/
     /*  SUPPORT FUNCTIONS TO HANDLE THE INSTAGRAM ACCESS TOKEN
     /*===================================================================*/

     private function getClientId() {
     return get_option('bean_inst_client_id');
     }

     /*===================================================================*/
     /*	SUPPORT FUNCTIONS TO HANDLE THE INSTAGRAM USER ID
     /*===================================================================*/

     private function getUserId() {
     return get_option('bean_inst_plugin_userid');
     }


     /*===================================================================*/
     /*	SUPPORT FUNCTIONS TO HANDLE THE INSTAGRAM POSTS CACHE
     /*===================================================================*/

     // GET THE INSTAGRAM POSTS CACHE
     private function getPostsCache() {
     return get_option('bean_inst_plugin_posts_' . $this->id);
     }

     // SET THE INSTAGRAM POSTS CACHE
     private function setPostsCache($cache) {
     update_option('bean_inst_plugin_posts_' . $this->id, $cache);
     }


     /*===================================================================*/
     /*	SUPPORT FUNCTIONS TO HANDLE THE INSTAGRAM FEED LAST CACHE TIME
     /*===================================================================*/

     // GET THE LAST CACHE TIME DEPENDING ON THE ID
     private function getLastCacheTime() {
     return get_option('bean_inst_plugin_last_cache_time_' . $this->id);
     }

     // SET THE LAST CACHE TIME DEPENDING ON THE ID
     private function setLastCacheTime($time) {
     update_option('bean_inst_plugin_last_cache_time_' . $this->id, $time);
     }

}

// A LITTLE WORK AROUND TO DELETE ALL THE OPTIONS ASSOCIATED WITH A WIDGET WHENEVER IT IS DELETED
function my_sidebar_admin_setup() {

	if ( 'post' == strtolower( $_SERVER['REQUEST_METHOD'] ) ) {

		$widget_id = $_POST['widget-id'];

		if ( isset( $_POST['delete_widget'] ) ) {
			if ( 1 === (int) $_POST['delete_widget'] ) {
				if ( strpos($widget_id, 'bean_instagram') !== FALSE ) {
					delete_option('bean_inst_plugin_accesstoken_' . $widget_id);
					delete_option('bean_inst_plugin_last_cache_time_' . $widget_id);
					delete_option('bean_inst_plugin_posts_' . $widget_id);
					delete_option('bean_inst_plugin_userid_' . $widget_id);
				}
			}
		}

	}

}





/**
 * Widget Settings Admin Page Output.
 * This section adds a "Bean Instagram" menu to the Settings dashboard link.
 *  
 *   
 * @package WordPress
 * @subpackage Bean Instagram
 * @author ThemeBeans
 * @since Bean Instagram 1.4
 */


/*===================================================================*/
/*  CREATE ADMIN LINK
/*===================================================================*/ 
function bean_instagram_options_page_settings() 
{
	add_options_page(
		__('Instagram Settings', 'bean'), __('Bean Instagram', 'bean'), 'manage_options', 'bean-instagram', 'bean_instagram_admin_page'
		);
} //END bean_instagram_options_page_settings

add_action( 'admin_menu', 'bean_instagram_options_page_settings' );


/*===================================================================*/
/*  LOAD ADMIN JS
/*===================================================================*/ 
function bean_instagram_enqueue_scripts($hook)  {

	if ('settings_page_bean-instagram' !== $hook) return;

	wp_enqueue_script( 'bean-instagram-script', BEAN_INSTAGRAM_PATH . 'js/bean-instagram.js', array(), '1.0', true );
} //END bean_instagram_enqueue_scripts

add_action( 'admin_enqueue_scripts', 'bean_instagram_enqueue_scripts' );


/*===================================================================*/
/*  REGISTER SETTINGS
/*===================================================================*/  
add_action('admin_init', 'bean_instagram_register_settings');

function bean_instagram_settings() {
	$bean_inst = array();
	$bean_inst[] = array('label' => 'Client ID:', 'name' => 'bean_inst_client_id');
	$bean_inst[] = array('label' => 'Client Secret:', 'name' => 'bean_inst_client_secret');
	$bean_inst[] = array('label' => 'Access Token:', 'name' => 'bean_inst_access_token', 'render' => false);
	$bean_inst[] = array('label' => 'User ID:', 'name' => 'bean_inst_plugin_userid', 'render' => false);

	return $bean_inst;
} //END bean_instagram_settings

function bean_instagram_register_settings() {
	$settings = bean_instagram_settings();
	foreach($settings as $setting) {
		register_setting('bean_instagram_settings', $setting['name'], isset($setting['sanitize_callback']) ? $setting['sanitize_callback'] : '');
	}
} //END bean_instagram_register_settings


/*===================================================================*/
/*  DELETE SETTINGS WHEN THE PLUGIN IS DEACTIVATED
/*===================================================================*/  

function bean_instagram_delete_plugin_options() {
	$settings = bean_instagram_settings();
	foreach($settings as $setting) {
		delete_option($setting['name']);
	}
}

register_deactivation_hook( BEAN_INSTAGRAM_PLUGIN_FILE, 'bean_instagram_delete_plugin_options' );


/*===================================================================*/
/*  CREATE THE SETTINGS PAGE
/*===================================================================*/  
function bean_instagram_admin_page($id) 
{
	if( !current_user_can('manage_options') ) { wp_die( __('Insufficient permissions', 'bean') ); }

	$settings = bean_instagram_settings();

	$settings_updated = isset($_GET['settings-updated']);

	$prefix = (!empty($_SERVER['HTTPS']) ? "https://" : "http://");
	$uri_parts = explode('?', $_SERVER['REQUEST_URI'], 2);
	$redirectURI = $prefix . $_SERVER['HTTP_HOST'] . $uri_parts[0] . '?page=bean-instagram';

	$code = isset($_GET['code']) ? $_GET['code'] : '';

    // Check if code parameter (from Instagram oauth) was passed and if so, get the access token from the Instagram API
	$code_valid = isset($_GET['code']) ? $_GET['code'] : '';

	if ($code_valid && !isset($_GET['settings-updated']) ) {
		$ch = curl_init();

		curl_setopt($ch, CURLOPT_URL, "https://api.instagram.com/oauth/access_token");
		curl_setopt($ch, CURLOPT_POSTFIELDS, "client_id=" . get_option('bean_inst_client_id') .
			"&client_secret=" . get_option('bean_inst_client_secret') .
			"&grant_type=authorization_code" .
			"&redirect_uri=" . urlencode($redirectURI) .
			"&code=" . $code );

		curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
		curl_setopt($ch, CURLOPT_REFERER, $redirectURI);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

		$result = curl_exec($ch);

		curl_close($ch);

	   // JSON decode the result
		$result = json_decode($result);

		if (isset($result->access_token)) {
			$accesstoken_value = $result->access_token;
			$userid_value = $result->user->id;
		}
	}

	echo '<div class="wrap">';
	screen_icon();
	echo '<h2>Bean Instagram Plugin</h2>';
	echo '<div class="wrap">'; 
	echo '<p>' . __('Display your Instagram photostream with the Bean Instagram plugin.  After entering your codes, simply add the Bean Instagram widget to a widget area within your Widgets Dashboard. If you need additional help, we wrote a <a href="http://themebeans.com/tutorials/retrieving-your-instagram-id-secret/" target="_blank">quick guide</a> to help you along.', 'bean' ) . '</p></br>';
	?>
	<?php
	echo '<form method="post" action="options.php">';



	echo '<h4 style="font-size: 15px; font-weight: 600; color: #222; margin-bottom: 10px;">' . __('How To', 'bean' ) . '</h4>';

	echo '<ol>';
	
	echo '<li>' . __( 'Visit the <strong><a href="http://instagram.com/developer" target="_blank">Instagram Developer</a></strong> portal and complete the following steps:', 'bean' ) . '</li>';

	/* translators: Click is used as a verb. */

	echo '<ol>';

	printf( '<li>' . __( 'Click the %1$s button on the top bar.', 'bean' ) . '</li>', "<strong>Manage Clients</strong>");

	/* translators: Click is used as a verb. */

	printf( '<li>' . __( 'Click the %1$s button.', 'bean' ) . '</li>', "<strong>Register a New Client</strong>");

	printf( '<li>' . __( 'Enter your application name, description and website.', 'bean' ) . '</li>');

	printf( '<li>' . __( 'Set the %1$s field to: %2$s', 'bean' ) . '</li>' , "<strong>OAuth redirect_uri</strong>", "<strong>$redirectURI</strong>");

	echo '<li>' . __( 'After all fields are complete, click the green <strong>Register</strong> button.', 'bean' ) . '</li>';

	echo '<li>' . __( 'Keep the Instagram Clients page open and head back to this page.', 'bean' ) . '</li>';

	echo '</ol></br>';

	printf( '<li>' . __( 'Copy & paste the %1$s and %2$s codes of your Instagram Client in the OAuth Codes form below.', 'bean' ) . '</li>', "<strong>Client ID</strong>", "<strong>Client Secret</strong>" );

	/* translators: Click the "Save Changes" button below. */

	printf( '<li>' . __( 'Click the %1$s button below.', 'bean' ) . '</li>', "<strong>" . __( 'Save Changes' ) . "</strong>" );

	/* translators: Click the "Retrieve Access Token" button below. */

	printf( '<li>' . __( 'Once the page refreshes, click the new %1$s button.', 'bean' ) . '</li>', '<strong>' . __( 'Retrieve Access Token', 'bean' ) . '</strong>' );

	/* translators: Click the "Save Changes" button below again. */

	printf( '<li>' . __( 'Once directed to authorize the request, click the green <strong>Authorize</strong> button. When authorization is granted, this page will refresh.', 'bean' ) . '</li>', '<strong>' . __( 'Retrieve Access Token', 'bean' ) . '</strong>' );

	printf( '<li>' . __( 'Click the blue %1$s button below the OAuth Codes one last time. Your access token should be "Available".', 'bean' ) . '</li>', "<strong>" . __( 'Save Changes' ) . "</strong>" );

	printf( '<li>' . __( 'Add the %1$s widget to a widget area in your %2$s.', 'bean' ) . '</li>', '<strong>Bean Instagram</strong>', '<a href="widgets.php"><strong>' . __( 'Widgets Dashboard' ) . '</strong></a>' );
	
	echo '</ol></br>';

	settings_fields('bean_instagram_settings');

	echo '<h4 style="font-size: 15px; font-weight: 600; color: #222; margin-bottom: 7px;">' . __('OAuth Codes', 'bean' ) . '</h4>';

	echo '<table>';
	foreach($settings as $setting) {
		if (isset($setting['render']) && $setting['render'] === false) continue;

		echo '<tr>';
		echo '<td style="padding-right: 20px;">' . $setting['label'] . '</td>';
		echo '<td><input type="text" style="width:500px;" name="'.$setting['name'].'" id="'.$setting['name'].'" value="'.get_option($setting['name']).'" onchange="clear_accesstoken(\'' . sprintf( __( 'Not available! Kindly click %1$s after entering the Client ID above.', 'bean'), __( 'Save Changes' ) ) . '\');"></td>';
		echo '</tr>';
	}

	echo '<tr>';
	echo '<td style="padding: 0 20px 0 0;">' . __('Access Token:', 'bean') . '</td>';
	echo '<td style="padding: 20px 0;">';

	$stored_accesstoken = get_option("bean_inst_access_token", "");
	$stored_userid = get_option("bean_inst_plugin_userid", "");

	if (empty($stored_accesstoken) && !isset($accesstoken_value)) {
		$client_id_value = get_option('bean_inst_client_id');

		if (!$code_valid && empty($client_id_value)) {
			echo '<strong style="color: red;">';
			printf( __( 'Not available! Kindly click %1$s after entering the Client ID above.', 'bean'), __( 'Save Changes' ) );
			echo '</strong>';
		} else {
			echo '<input type="button" class="button" onclick="bean_instagram_get_access_token.call(this, \'' . $redirectURI . '\'); return false;" value="' . __( 'Get the Access Token', 'bean' ) . '">';

			if ($code_valid && !isset($accesstoken_value)) {
				echo "<strong style='padding-left: 10px; line-height: 28px;'>";
				echo __("There was some error while trying to retrieve the access token. Make sure that the Redirect URI in your Instagram API Client is set to exactly what is suggested above.", 'bean');
				echo "</strong>";
			}
		}

	} else {
		echo '<strong id="bean_inst_access_token_status" style="color: #FFF; background-color: #2BC28B; border-radius: 2px; padding: 3px 6px;">Available';

		if (empty($stored_accesstoken)) {
			echo ' – ';
			echo __('Click "Save Changes" below to save the Access Token. (It will be lost otherwise.)', 'bean');
		}

		echo '</strong>';
	}

	echo '<input type="hidden" id="bean_inst_access_token" name="bean_inst_access_token" value="' . (isset($accesstoken_value) ? $accesstoken_value : $stored_accesstoken).'">';

	echo '<input type="hidden" id="bean_inst_plugin_userid" name="bean_inst_plugin_userid" value="' . (isset($userid_value) ? $userid_value : $stored_userid).'">';

	echo '</td>';
	echo '</tr>';

	echo '</table>';

	submit_button();

	echo '</form>';
	echo '</div>';
	echo '</div>';
} //END bean_instagram_admin_page


?>